### DISCUSSION TITLE
Automatic module import isn't working correctly

### DISCUSSION THREAD

**Discussion by chscott** [PowerShell/PowerShell#17854] [General]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I have a custom (proprietary) module that gets installed as part of every AWS instance we create. When I launch PowerShell and attempt to run a command _for the first time_, the command is found well enough to tab-complete the parameters. I don't experience any problem there. However, when I try to run the command, I get a sequence like this:

```
PS C:\Windows\System32> Build-ArtifactArchive -ArtifactType Customer -Verbose
Build-ArtifactArchive: The term 'Build-ArtifactArchive' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
PS C:\Windows\System32> Get-Module

ModuleType Version    PreRelease Name                                ExportedCommands
---------- -------    ---------- ----                                ----------------
Manifest   7.0.0.0               Microsoft.PowerShell.Management     {Add-Content, Clear-Content, Clear-Item, Clear-ItemProperty…}
Binary     2.1.0.0               Microsoft.PowerShell.PSReadLine2    {Get-PSReadLineOption, Set-PSReadLineOption, Set-PSReadLineKeyHandler, Get-PSReadLineKeyHandler…}
Manifest   7.0.0.0               Microsoft.PowerShell.Utility        {Add-Member, Add-Type, Clear-Variable, Compare-Object…}
Script     2.1.0                 PSReadLine                          {Get-PSReadLineKeyHandler, Get-PSReadLineOption, Remove-PSReadLineKeyHandler, Set-PSReadLineKeyHandle…
Script     1.0.1353.0            TMCTools                            {Build-ArtifactArchive, Build-ExceptionReport, Build-FolderReport, Build-JobReport…}
Binary     1.0.1353              TMCTools

PS C:\Windows\System32> Remove-Module TMCTools
PS C:\Windows\System32> Build-ArtifactArchive -ArtifactType Customer -Verbose
2022-08-05_09:49:56 No license key was detected.
Exception: Unable to initialize function
```

As you can see, the module is loaded in the session and the command is recognized as an exported command, yet it can't be found. But then simply removing the module from the session and repeating the command (this autoloads it again, right?) works.

Note that the module appears as both Script and Binary. This is because I use PSProtector to compile it to DLL.

### Expected behavior

```console
The command should work on the first invocation via autoload.
```

### Actual behavior

```console
PowerShell says the command cannot be found even though Get-Module shows it is loaded.
```

### Environment data

```powershell
PS C:\Windows\System32> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.5
PSEdition                      Core
GitCommitId                    7.2.5
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by fMichaleczek** [comment]
@chscott I think it's out-of-the scope of an issue for the source code of PowerShell, the problem is on the side of the module TMCTools. You should move it to discussions or ask help on stackoverflow

**Comment by chscott** [comment]
> @chscott I think it's out-of-the scope of an issue for the source code of PowerShell, the problem is on the side of the module TMCTools. You should move it to discussions or ask help on stackoverflow

How is it a module issue?

**Comment by fMichaleczek** [comment]
> > @chscott I think it's out-of-the scope of an issue for the source code of PowerShell, the problem is on the side of the module TMCTools. You should move it to discussions or ask help on stackoverflow
> 
> How is it a module issue?

There is the same module loaded with 2 versions. I think there is a problem on the psd1

**Comment by chscott** [comment]
As I said, it's not really two versions. It just appears that way because I use PSProtector. I'm not saying that's definitely not the issue, but this has never been a problem for me until recently, though I couldn't tell you the last version of PowerShell where this wasn't a problem. Is there a way to trace the module loading bit? I thought -Verbose might do it, but I don't get anything there.

**Reply by fMichaleczek** [reply]
With the PSProtector demo, I create a _test_ module composed with a single command _Get-TestCommand_

As you see, PSProtector replace Get-Command and Get-Module. I will not do more reverse-enginerring, but I hope you see the point about having weird symptoms (I see your previous issue after)

![image](https://user-images.githubusercontent.com/6187529/183146343-d3395d65-1a5c-44c5-a26f-20cc57cc0c96.png)

Module Autoloading is not working. And it's normal, because there is not property ExportedCommands on the Module.

![image](https://user-images.githubusercontent.com/6187529/183149249-13a0085b-e345-49a8-9e2a-cd5af2e237ee.png)

At startup (and later), PowerShell scan all available Modules and search for the property "ExportedCommands" to make a cache which will be the base of the autoloading.

I suggest you to read the PSModuleAnalysisCachePath section of this link, and clear your cache (or disabled it on your devbox)
I think your original PSM1 is in it.

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.2#powershells-environment-variables

![image](https://user-images.githubusercontent.com/6187529/183153072-6426b596-e1f2-44a5-b414-4c5132ca9f57.png)

**Reply by chscott** [reply]
What you describe is not what I'm seeing, though.

<img width="911" alt="image" src="https://user-images.githubusercontent.com/9805298/183671819-b571e10a-651c-4cd7-9fc9-64696bdfebd3.png">

Doesn't this seem like a problem with PowerShell and autoloading?

**Reply by chscott** [reply]
Also, I can see the impacted command in the module cache.

<img width="532" alt="image" src="https://user-images.githubusercontent.com/9805298/183673247-c44b21f3-0cc1-466c-9f67-98326982e626.png">

**Comment by iSazonov** [comment]
> Is there a way to trace the module loading bit?

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/trace-command?view=powershell-7.2
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/get-tracesource?view=powershell-7.2

<blockquote><img src="https://docs.microsoft.com/media/logos/logo-powershell-social.png" width="48" align="right"><div><strong><a href="https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/trace-command">Trace-Command (Microsoft.PowerShell.Utility) - PowerShell</a></strong></div><div>The Trace-Command cmdlet configures and starts a trace of the specified expression or command. It works like Set-TraceSource, except that it applies only to the specified command. </div></blockquote>
<blockquote><img src="https://docs.microsoft.com/media/logos/logo-powershell-social.png" width="48" align="right"><div><strong><a href="https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/get-tracesource">Get-TraceSource (Microsoft.PowerShell.Utility) - PowerShell</a></strong></div><div>The Get-TraceSource cmdlet gets the trace sources for PowerShell components that are currently in use. You can use the data to determine which PowerShell components you can trace. When tracing, the component generates detailed messages about each step in its internal processing. Developers use the trace data to monitor data flow, program execution, and errors. The tracing cmdlets were designed for PowerShell developers, but they are available to all users. </div></blockquote>

**Comment by kasini3000** [comment]
I love this topic,
It looks like the problem is in your module, does this solve it?
```
Remove-Module TMCTools
start-sleep -second 1
import-Module TMCTools
```

**Reply by chscott** [reply]
That works, but the `Start-Sleep` isn't necessary. If you look at the original post, this (simplified) sequence is all it takes:

1. Run a command exported by the module
2. `Remove-Module`
3. Run the same command as in Step 1

**Comment by doctordns** [comment]
Is the module in question a WIndows PowerShell 5.1 module that PowerShell 7 loads via the compatibility mechanism?
